### PR TITLE
Kafka plugin: Support TLS

### LINKED ISSUE #141: Kafka plugin: Use TLS certificates when communication with Kafka broker
### Feature description

The Kafka plugin should be able to securely communicate with Kafka broker using TLS certificates.

Depends on #137.

### PR DESCRIPTION
**Pull request by hariso**
### Description

Adds TLS support in the Kafka plugin. The changes (as they are now) expect the needed files to be present on the machine where Conduit is running.

Fixes #141 

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

**Comment by lovromazgon** [comment]
@hariso I propose moving this PR to the extracted [kafka plugin](https://github.com/ConduitIO/conduit-plugin-kafka), since we will move to that soon and it's not a feature that's urgently needed at this moment. WDYT?

**Review by lovromazgon — COMMENTED** [review]
More or less just nitpicks, looks good overall 👍 As I said in the previous comment, I think we should add the change to the other repository rather than here, it would probably create conflicts with the conduit-plugin PR, so additional work that doesn't bring us any value.

File: pkg/plugins/kafka/config.go [hunk 1]
```diff
@@ -88,9 +97,39 @@ func Parse(cfg map[string]string) (Config, error) {
...
+	// Parse InsecureSkipVerify, default is 'false'
+	insecureString, ok := cfg[InsecureSkipVerify]
+	if ok {
+		insecure, err := strconv.ParseBool(insecureString)
+		if err != nil {
+			return cerrors.Errorf("value %q for InsecureSkipVerify is not valid", insecureString)
+		}
+		parsed.InsecureSkipVerify = insecure
+	} else {
+		parsed.InsecureSkipVerify = false
```
[the remark this answers is not in the document]
Comment (hariso) [reply pkg/plugins/kafka/config.go:128]: Good point, I'll remove this part.

File: pkg/plugins/kafka/consumer.go [hunk 2]
```diff
@@ -88,7 +94,45 @@ func newReader(cfg Config, groupID string) *kafka.Reader {
...
+		dialer, err := newTLSDialer(cfg)
+		if err != nil {
+			return nil, cerrors.Errorf("couldn't create dialer: %w", err)
+		}
+		readerCfg.Dialer = dialer
+	}
+	return kafka.NewReader(readerCfg), nil
+}
+
+func newTLSDialer(cfg Config) (*kafka.Dialer, error) {
```
[the remark this answers is not in the document]
Comment (hariso) [reply pkg/plugins/kafka/consumer.go:108]: Yup, makes sense!

File: pkg/plugins/kafka/consumer.go [hunk 3]
```diff
@@ -88,7 +94,45 @@ func newReader(cfg Config, groupID string) *kafka.Reader {
...
+		return nil, cerrors.Errorf("invalid TLS config: %w", err)
+	}
+	return &kafka.Dialer{
+		ClientID:  "",
+		DualStack: true,
+		TLS:       tlsCfg,
+	}, nil
+}
+
+func newTLSConfig(clientCert, clientKey, caCert string, serverNoVerify bool) (*tls.Config, error) {
```
[the remark this answers is not in the document]
Comment (hariso) [reply pkg/plugins/kafka/consumer.go:120]: 👍

File: pkg/plugins/kafka/config.go [hunk 4]
```diff
@@ -88,9 +97,39 @@ func Parse(cfg map[string]string) (Config, error) {
...
 }
 
+func setTLSConfigs(parsed *Config, cfg map[string]string) error {
+	// All three values should be set so that TLS works
+	// If none of the three values are set, then TLS should not be used.
+	tlsCfgOk := (cfg[ClientCert] == "") == (cfg[ClientKey] == "")
+	tlsCfgOk = tlsCfgOk && (cfg[ClientKey] == "") == (cfg[CACert] == "")
+	if !tlsCfgOk {
+		return cerrors.New("TLS config not OK (all values need to be set or all values need to be unset)")
+	}
```
[the remark this answers is not in the document]
Comment (hariso) [reply pkg/plugins/kafka/config.go:115]: You have a point here. I'm discussing with Lyubo whether mutual TLS is required, so based on that I'll  update this code. Thanks for the suggestion!

**Comment by hariso** [comment]
Closing in favor of https://github.com/ConduitIO/conduit-connector-kafka/pull/3

**Pull request closed without merging**